Kernel/Storage/NVMe/NVMeController.cpp

```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright (c) 2021, Pankaj R <pankydev8@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "NVMeController.h"
+#include "AK/Format.h"
+#include <AK/RefPtr.h>
+#include <AK/Types.h>
+#include <Kernel/Arch/x86/IO.h>
+#include <Kernel/Arch/x86/Processor.h>
+#include <Kernel/Bus/PCI/API.h>
+#include <Kernel/Devices/Device.h>
+#include <Kernel/FileSystem/ProcFS.h>
+#include <Kernel/Sections.h>
+
+namespace Kernel {
+Atomic<u8> NVMeController::controller_id {};
+
+ErrorOr<NonnullRefPtr<NVMeController>> NVMeController::try_initialize(const Kernel::PCI::DeviceIdentifier& device_identifier)
+{
+    auto controller = TRY(adopt_nonnull_ref_or_enomem(new NVMeController(device_identifier)));
+    TRY(controller->initialize());
+    NVMeController::controller_id++;
+    return controller;
+}
+
+NVMeController::NVMeController(const PCI::DeviceIdentifier& device_identifier)
+    : PCI::Device(device_identifier.address())
+    , m_pci_device_id(device_identifier)
+{
+}
+
+ErrorOr<void> NVMeController::initialize()
+{
+    // Nr of queues = one queue per core
+    auto nr_of_queues = Processor::count();
+    auto irq = m_pci_device_id.interrupt_line().value();
+
+    PCI::enable_memory_space(m_pci_device_id.address());
+    PCI::enable_bus_mastering(m_pci_device_id.address());
+    auto bar = PCI::get_BAR0(m_pci_device_id.address()) & 0xFFFFFFF0;
+    auto bar_len = round_up_to_power_of_two(CTRL_REG_SIZE(nr_of_queues), 4096);
+    auto region_or_error = TRY(MM.allocate_kernel_region(PhysicalAddress(bar), bar_len, "PCI NVMe BAR", Memory::Region::Access::ReadWrite));
+    // Queues need access to the doorbell registers, hence leaking the pointer
+    m_controller_regs = region_or_error.leak_ptr();
```

Comment:
Please don't leak a pointer here. Maybe using `TypedMapping` struct more could help you to do this more cleanly :)